api/src/main/java/org/apache/iceberg/expressions/Binder.java

```diff
@@ -91,12 +91,21 @@ static Expression bind(StructType struct,
   }
 
   public static Set<Integer> boundReferences(StructType struct, List<Expression> exprs, boolean caseSensitive) {
+    return references(struct, exprs, caseSensitive, false);
+  }
+
+  public static Set<Integer> references(
+      StructType struct, List<Expression> exprs, boolean caseSensitive, boolean alreadyBound) {
```

Comment:
Can we detect that an expression is unbound? Maybe identify named refs and return? Then we could just use one method for everything.

Reply:
Can i just use `instanceof Unbound` to detect if an expression is unbound? Something like this:

```
  private static boolean isUnbound(Expression expr) {
    switch (expr.op()) {
      case TRUE:
        return false;
      case FALSE:
        return false;
      case NOT:
        Not not = (Not) expr;
        return isUnbound(not.child());
      case AND:
        And and = (And) expr;
        return isUnbound(and.left()) || isUnbound(and.right());
      case OR:
        Or or = (Or) expr;
        return isUnbound(or.left()) || isUnbound(or.right());
      default:
        return expr instanceof Unbound;
    }
  }
```
Then the method `boundReferences` can be
```
  public static Set<Integer> boundReferences(StructType struct, List<Expression> exprs, boolean caseSensitive) {
    if (exprs == null) {
      return ImmutableSet.of();
    }
    ReferenceVisitor visitor = new ReferenceVisitor();
    for (Expression expr : exprs) {
      if (isUnbound(expr)) {
        ExpressionVisitors.visit(bind(struct, expr, caseSensitive), visitor);
      } else {
        ExpressionVisitors.visit(expr, visitor);
      }
    }
    return visitor.references;
  }
```

Reply:
Sounds good to me!